### PR TITLE
LibWeb: Add basic HTML meter element support

### PR DESCRIPTION
**Pull request by bplaat**
Adding support for a new HTML form element is always nice, so I have added the support for the `<meter>` element. Most things work, I used the `-webkit-` style pseudo elements to implement the bars so you can style the with CSS. This is mostly copied from the `::placeholder` part of the `input` element.

Screenshot 2023-11-15 at 19 55 11
*Screenshot of the meters test page*

### CODE REVIEW

**Comment by AtkinsSJ** [comment]
Drive-by phone review: There are html-specific float parsing functions for use in attributes. I don't remember where they are though. 😅 The spec usually mentions the algorithm in the attribute description, though I didn't spot it when I had a quick look just now.

**Comment by bplaat** [comment]
There is `HTML/NUmber.cpp` with `parse_integer` and `parse_non_negative_integer`, no floating point parsing function but I will add one and apply them where needed in the codebase.

File: Userland/Libraries/LibWeb/HTML/Numbers.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
  */
 
 #include <AK/GenericLexer.h>
+#include <LibC/math.h>
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/Numbers.cpp:8]: Nit-pick: I think this should just be `<math.h>`?

File: Userland/Libraries/LibWeb/CSS/Default.css [hunk 2]
```diff
@@ -787,10 +810,10 @@ progress {
...
     height: inherit;
-    background-color: grey;
+    background-color: lightgray;
 }
 
 progress::-webkit-progress-value {
     height: inherit;
-    background-color: green;
+    background-color: skyblue;
 }
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Default.css:819]: This is unrelated to adding `<meter>`

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp [hunk 3]
```diff
@@ -370,15 +371,68 @@ ErrorOr<void> TreeBuilder::create_layout_tree(DOM::Node& dom_node, TreeBuilder::
         pop_parent();
     }
 
+    if (is<HTML::HTMLMeterElement>(dom_node)) {
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:374]: Constructing the inner parts of this in TreeBuilder feels like the wrong approach to me. I think it should be done using a shadow tree inside the HTMLMeterElement. For an example, see [`HTMLInputElement`](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp#L535).

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Always cool to see a new HTML element implemented! :tada:

It does seem to be malfunctioning a little though. Here's how it looks for me, with this branch based on current master:
image

Also, from looking in the inspector, the pseudo-elements aren't appearing for many of the `<meter>`.

**Comment by bplaat** [comment]
@AtkinsSJ sorry for the long wait but I was quite busy. I have fixed all your complaints and now the `<meter>` element uses a shadow dom tree like the input element. This system works quite nice, but the faking of a div element that uses a CSS pseudo selector is very hacky :^)

I saw in Google Chrome that they have a custom `psuedo="-webkit-meter-optimum-value"` attribute when using the shadow dom. Maybe we can also add support for such attribute then the custom `PlaceholderElement, `MeterBarElement` and `MeterValueElement` can be removed.

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp [hunk 4]
```diff
@@ -402,8 +402,7 @@ ErrorOr<void> TreeBuilder::create_layout_tree(DOM::Node& dom_node, TreeBuilder::
             bar_style->set_property(CSS::PropertyID::Display, CSS::DisplayStyleValue::create(CSS::Display::from_short(CSS::Display::Short::FlowRoot)));
             auto value_style = TRY(style_computer.compute_style(progress, CSS::Selector::PseudoElement::ProgressValue));
             value_style->set_property(CSS::PropertyID::Display, CSS::DisplayStyleValue::create(CSS::Display::from_short(CSS::Display::Short::Block)));
-            auto position = progress.position();
-            value_style->set_property(CSS::PropertyID::Width, CSS::PercentageStyleValue::create(CSS::Percentage(position >= 0 ? round_to<int>(100 * position) : 0)));
+            value_style->set_property(CSS::PropertyID::Width, CSS::PercentageStyleValue::create(CSS::Percentage(progress.position() * 100)));
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:411]: Is this in the wrong commit maybe?

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp [hunk 5]
```diff
@@ -22,4 +26,185 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
...
+
+// https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-actual
+double HTMLMeterElement::value() const
+{
+    auto maybe_value_string = get_attribute(HTML::AttributeNames::value);
+    if (!maybe_value_string.has_value())
+        return 0;
+    auto maybe_value = parse_floating_point_number(maybe_value_string.value());
+    if (!maybe_value.has_value())
+        return 0;
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp:43]: The spec describes using 0 as the "candidate actual value", which is then clamped between min and max. So we can't just return 0 here.

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp [hunk 6]
```diff
@@ -22,4 +26,185 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
...
+
+// https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-maximum
+double HTMLMeterElement::max() const
+{
+    auto maybe_max_string = get_attribute(HTML::AttributeNames::max);
+    if (!maybe_max_string.has_value())
+        return 1;
+    auto maybe_max = parse_floating_point_number(maybe_max_string.value());
+    if (!maybe_max.has_value())
+        return 1;
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp:81]: Similar to my comment on `value()`: 1 should be the "candidate" value which we then ensure is >= min.

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp [hunk 7]
```diff
@@ -22,4 +26,185 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
...
+}
+
+void HTMLMeterElement::update_meter_value_element()
+{
+    if (!m_meter_value_element)
+        return;
+
+    // UA requirements for regions of the gauge:
+    // If the optimum point is equal to the low boundary or the high boundary, or anywhere in between them, then the region between the low and high boundaries of the gauge must be treated as the optimum region, and the low and high parts, if any, must be treated as suboptimal.
+    if (optimum() >= low() && optimum() <= high()) {
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp:181]: Each of these methods does quite a lot of logic, including parsing numbers. I think it would be worth calling them once each, saving their values, and then comparing those instead.

(Or as a step further, maybe it's worth caching these float values as `m_value`, `m_low`, etc, and setting those when the attribute changes? A lot of the getter methods call the other getter methods so things get parsed multiple times anyway.)

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp [hunk 8]
```diff
@@ -22,4 +26,185 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
...
+        if (value() >= low() && value() <= high())
+            m_meter_value_element->set_pseudo_element(CSS::Selector::PseudoElement::MeterOptimumValue);
+        else
+            m_meter_value_element->set_pseudo_element(CSS::Selector::PseudoElement::MeterSuboptimumValue);
+    }
+    // Otherwise, if the optimum point is less than the low boundary, then the region between the minimum value and the low boundary must be treated as the optimum region, the region from the low boundary up to the high boundary must be treated as a suboptimal region, and the remaining region must be treated as an even less good region.
+    else if (optimum() < low()) {
+        if (value() < low())
+            m_meter_value_element->set_pseudo_element(CSS::Selector::PseudoElement::MeterOptimumValue);
+        else if (value() >= low() && value() <= high())
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp:191]: `value() >= low()` is always true here.

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp [hunk 9]
```diff
@@ -22,4 +26,185 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
...
+            m_meter_value_element->set_pseudo_element(CSS::Selector::PseudoElement::MeterSuboptimumValue);
+        else
+            m_meter_value_element->set_pseudo_element(CSS::Selector::PseudoElement::MeterEvenLessGoodValue);
+    }
+    // Finally, if the optimum point is higher than the high boundary, then the situation is reversed; the region between the high boundary and the maximum value must be treated as the optimum region, the region from the high boundary down to the low boundary must be treated as a suboptimal region, and the remaining region must be treated as an even less good region.
+    else {
+        if (value() > high())
+            m_meter_value_element->set_pseudo_element(CSS::Selector::PseudoElement::MeterOptimumValue);
+        else if (value() >= low() && value() <= high())
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp:200]: `value() <= high()` is always true here.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Playing around with this now, it works really well. Thanks! :smiley:

I did spot a couple of issues with `HTMLMeterElement::value()` and friends. The spec does have clear algorithm-y steps for these, so IMO it would be worth adding the spec comments and following those directly.

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp (hunk 4 above)
Comment (bplaat) [reply Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:411]: Nope I changed the float parsing code of progressbar in the same commit

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp (hunk 7 above)
Comment (bplaat) [reply Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp]: I worry about desync issues so for now I query the attrs once as you suggested

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp [hunk 10]
```diff
@@ -24,4 +28,219 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
...
+    if (maybe_high_string.has_value()) {
+        auto maybe_high = parse_floating_point_number(maybe_high_string.value());
+        if (maybe_high.has_value())
+            candidate_high = maybe_high.value();
+    }
+
+    // If the candidate high boundary is less than the low boundary, then the high boundary is the low boundary.
+    // Otherwise, if the candidate high boundary is greater than the maximum value, then the high boundary is the maximum value.
+    // Otherwise, the high boundary is the candidate high boundary.
+    return clamp(candidate_high, min(), max());
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp:149]: This wants to be clamped between `low()` and `max()`.

File: Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp [hunk 11]
```diff
@@ -24,4 +28,219 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
...
+    if (!m_meter_value_element)
+        return;
+
+    // UA requirements for regions of the gauge:
+    double m_value = value();
+    double m_min = min();
+    double m_max = max();
+    double m_low = low();
+    double m_high = high();
+    double m_optimum = optimum();
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp:218]: Naming these `m_foo` when they're not member variables is a bit odd. I get that's it's hard to find a good name though. Could always do `double foo = this->foo();` if nothing else.

File: Userland/Libraries/LibWeb/CSS/Default.css [hunk 12]
```diff
@@ -67,6 +67,30 @@ option {
...
+    width: 300px;
+    height: 12px;
+}
+meter::-webkit-meter-bar, meter::-webkit-meter-optimum-value, meter::-webkit-meter-suboptimum-value, meter::-webkit-meter-even-less-good-value {
+    display: block;
+    height: 100%;
+}
+meter::-webkit-meter-bar {
+    background-color: hsl(0, 0%, 96%);
+    outline: 1px solid rgba(0, 0, 0, 0.5);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Default.css:82]: Why use an outline instead of a border here?

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Thanks! Including the spec comments also helped me find a little bug so it's already paying off. :wink:

If possible it would be good to get some tests added for `<meter>` to make sure nothing breaks in the future.. Mostly a layout test would do, but for the different "optimum/suboptimum/etc" states I think it would have to be a ref test.

**Review by AtkinsSJ — APPROVED** [review]
Let's gooooo!
